### ISSUE TITLE
Native: getLink() on SharedLibraryBinarySpec.TasksCollection and NativeExecutableBinarySpec.TasksCollection returns null

### ISSUE THREAD

**Issue by philmcardle** [gradle/gradle#817]
### Expected Behavior
https://docs.gradle.org/current/javadoc/org/gradle/nativeplatform/SharedLibraryBinarySpec.TasksCollection.html
https://docs.gradle.org/current/javadoc/org/gradle/nativeplatform/NativeExecutableBinarySpec.TasksCollection.html

These suggest that I should be able to call getLink() and retrieve the link Task.

### Current Behavior
When I call getLink() on instances of these objects, I get null in return.

Contrarily, if I iterate over the same tasks collection, I will find a task of type LinkExecutable_Decorated or LinkSharedLibrary_Decorated depending on which collection I'm using.

### Context
I was trying to retrieve the link task for a NativeExecutableSpec using this method, and have to search the tasks collection instead.

### Steps to Reproduce (for bugs)
Using the [cpp-exe](https://github.com/gradle/gradle/tree/master/subprojects/docs/src/samples/native-binaries/cpp-exe) or [cpp-lib](https://github.com/gradle/gradle/tree/master/subprojects/docs/src/samples/native-binaries/cpp-lib) samples, attempt to call `tasks.getLink()` inside `binaries.all {` (cpp-exe) or `binaries.withType(SharedLibraryBinarySpec) {` (cpp-lib)

Call any gradle task (tasks/build/assemble, etc.) to see that the output of `getLink()` is null.

See also that in the cpp-exe sample, iterating `tasks.all` shows a task of class `LinkExecutable_Decorated` with name `linkMainExecutable`, with similar behaviour in the cpp-lib sample.

The same also appears to be true of `getInstall()`, though I've only tested that lightly while reporting this bug.

### Your Environment
Windows 10 1607 with MSVC 2013 & MSVC 2015, Gradle 3.1
Debian Jessie with gcc 4.9.2, Gradle 3.1

**Comment by philmcardle** [comment]
I'm hoping I'm doing something wrong, because I'm seeing the same result calling getBuild() on NativeExecutableBinarySpec.TasksCollection

**Comment by philmcardle** [comment]
Okay, I tried `tasks.withType(org.gradle.nativeplatform.tasks.LinkExecutable)` which returns an empty set, at which point I figure I'm falling foul of a concurrency issue, and right enough, `tasks.withType(LinkExecutable).all` gives me a match, but `.each` does not.

Is there any point in the configuration at which getLink/getInstall/getBuild will be valid?

**Comment by lacasseio** [comment]
Thanks @philmcardle for raising this issue. I did encounter a similar issue in the past, we ended up refactoring the code a bit and using a workaround. I suspect this is caused by the order in which the task is created regarding the software model rule. This may be a limitation to the current software model. The `Task`s are created once the binary container is closed. But the current configuration is happening while the binary container is mutated. We are presented with a circular dependency. This will require more time to debug.

Keep in mind that `NativeBinary` can be mutated once the container is closed as those are `Unmanaged` type. However, this is very bad practice and will most likely result in failure for future versions of Gradle.
